File: app/src/main/java/de/christinecoenen/code/zapp/app/livestream/model/LiveShow.java
package de.christinecoenen.code.zapp.app.livestream.model;

import androidx.annotation.NonNull;

import org.joda.time.DateTime;
import org.joda.time.Duration;

public class LiveShow {

	private String title;
	private String subtitle;
	private String description;
	private DateTime startTime;
	private DateTime endTime;

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public String getSubtitle() {
		return subtitle;
	}

	public void setSubtitle(String subtitle) {
		this.subtitle = subtitle;
	}

	@SuppressWarnings("unused")
	public void setDescription(String description) {
		this.description = description;
	}

	public DateTime getStartTime() {
		return startTime;
	}

	public void setStartTime(DateTime startTime) {
		this.startTime = startTime;
	}

	public DateTime getEndTime() {
		return endTime;
	}

	public void setEndTime(DateTime endTime) {
		this.endTime = endTime;
	}

	public boolean hasDuration() {
		return startTime != null && endTime != null;
	}

	public float getProgressPercent() {
		Duration showDuration = new Duration(startTime, endTime);
		Duration runningDuration = new Duration(startTime, DateTime.now());
		return (float) runningDuration.getStandardSeconds() / showDuration.getStandardSeconds();
	}

	@NonNull
	@Override
	public String toString() {
		return "Show{" +
				"title='" + title + '\'' +
				", subtitle='" + subtitle + '\'' +
				", description='" + description + '\'' +
				", startTime=" + startTime +
				", endTime=" + endTime +
				'}';
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/livestream/model/LiveShow.kt
package de.christinecoenen.code.zapp.app.livestream.model

import android.content.Context
import de.christinecoenen.code.zapp.R
import org.joda.time.DateTime
import org.joda.time.Duration
import org.joda.time.format.DateTimeFormat
import org.koin.android.ext.koin.androidContext

data class LiveShow(
	var title: String,
	var subtitle: String? = null,
	var description: String? = null,
	var startTime: DateTime? = null,
	var endTime: DateTime? = null
) {

	val progressPercent: Float
		get() {
			val showDuration = Duration(startTime, endTime)
			val runningDuration = Duration(startTime, DateTime.now())
			return runningDuration.standardSeconds.toFloat() / showDuration.standardSeconds
		}

	val hasDuration
		get() = startTime != null && endTime != null

	fun formattedDuration(context: Context): String? {
		return if (hasDuration) {
			val startTime = startTime!!.toString(DateTimeFormat.shortTime())
			val endTime = endTime!!.toString(DateTimeFormat.shortTime())
			context.getString(R.string.view_program_info_show_time, startTime, endTime)
		} else {
			null
		}
	}

	companion object {
		fun getEmpty(context: Context): LiveShow {
			return LiveShow(context.getString(R.string.activity_channel_detail_info_error))
		}
	}
}
